File: 1071-greatest-common-divisor-of-strings/1071-greatest-common-divisor-of-strings.java
class Solution {
    public String gcdOfStrings(String str1, String str2) {
        int s1Len = str1.length();
        int s2Len = str2.length();
        
        if (!(str1 + str2).equals(str2 + str1)) {
            return "";
        } else if (s1Len == s2Len && str1.equals(str2)) {
            return str2;
        } else {
            // write code for all 'normal' test cases
            int minLen = Math.min(s1Len, s2Len);
            
            int GCD = 0;
            
            for (int i = 1; i <= minLen; i++) {
                if (s1Len % i == 0 && s2Len % i == 0) {
                    GCD = i;
                }
            }
            
            return str2.substring(0, GCD);
        }
    }
}

File: 1071-greatest-common-divisor-of-strings/1071-greatest-common-divisor-of-strings.js
/**
 * @param {string} str1
 * @param {string} str2
 * @return {string}
 */
var gcdOfStrings = function(str1, str2) {
    var str1Length = str1.length;
    var str2Length = str2.length;
    
    if (str1 + str2 !== str2 + str1) {
        return "";
    } else if (str2 === str1) {
        return str2;
    } else {
        var gcd = 0;
        var minLength = Math.min(str1.length, str2.length);
        for (let i = 1; i <= minLength; i++) {
            if (i !== 0 && str2.length % i === 0 && str1.length % i === 0) {
                gcd = i;
            }
        }
        return str2.substring(0, gcd);
    }
};